tests/src/tests_kalman_soc.cpp

```diff
@@ -0,0 +1,903 @@
+/*
+ * Copyright (c) The Libre Solar Project Contributors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "tests.h"
+#include <setup.h>
+#include <stdio.h>
+#include <time.h>
+
+#include "bat_charger.h"
+#include <math.h>
+#include <stdlib.h>
+#include <string.h>
+#include <strings.h>
+
+//#define DEBUG 0
+
+// GPS Model Backtest of TinyEKF Lib
+// See for implemented test example:
+// https://github.com/simondlevy/TinyEKF/tree/master/extras/c
+
+/* states */
+#define NUMBER_OF_STATES_GPS 8
+
+/* observables */
+#define NUMBER_OF_OBSERVABLES_GPS 4
+
+typedef struct
+{
+
+    int n; /* number of state values */
+    int m; /* number of observables */
+
+    float x[NUMBER_OF_STATES_GPS]; /* state vector */
+
+    float P[NUMBER_OF_STATES_GPS][NUMBER_OF_STATES_GPS]; /* prediction error covariance */
+    float Q[NUMBER_OF_STATES_GPS][NUMBER_OF_STATES_GPS]; /* process noise covariance */
+    float R[NUMBER_OF_OBSERVABLES_GPS]
+           [NUMBER_OF_OBSERVABLES_GPS]; /* measurement error covariance */
+
+    float G[NUMBER_OF_STATES_GPS][NUMBER_OF_OBSERVABLES_GPS]; /* Kalman gain; a.k.a. K */
+
+    float F[NUMBER_OF_STATES_GPS][NUMBER_OF_STATES_GPS];      /* Jacobian of process model */
+    float H[NUMBER_OF_OBSERVABLES_GPS][NUMBER_OF_STATES_GPS]; /* Jacobian of measurement model */
+
+    float Ht[NUMBER_OF_STATES_GPS]
+            [NUMBER_OF_OBSERVABLES_GPS];                  /* transpose of measurement Jacobian */
+    float Ft[NUMBER_OF_STATES_GPS][NUMBER_OF_STATES_GPS]; /* transpose of process Jacobian */
+    float Pp[NUMBER_OF_STATES_GPS][NUMBER_OF_STATES_GPS]; /* P, post-prediction, pre-update */
+
+    float fx[NUMBER_OF_STATES_GPS];      /* output of user defined f() state-transition function */
+    float hx[NUMBER_OF_OBSERVABLES_GPS]; /* output of user defined h() measurement function */
+
+    /* temporary storage */
+    float tmp0[NUMBER_OF_STATES_GPS][NUMBER_OF_STATES_GPS];
+    float tmp1[NUMBER_OF_STATES_GPS][NUMBER_OF_OBSERVABLES_GPS];
+    float tmp2[NUMBER_OF_OBSERVABLES_GPS][NUMBER_OF_STATES_GPS];
+    float tmp3[NUMBER_OF_OBSERVABLES_GPS][NUMBER_OF_OBSERVABLES_GPS];
+    float tmp4[NUMBER_OF_OBSERVABLES_GPS][NUMBER_OF_OBSERVABLES_GPS];
+    float tmp5[NUMBER_OF_OBSERVABLES_GPS];
+
+} ekf_gps_t;
```

Comment:
Is this struct not similar to one of the large structs previously defined? Could it be reused?

Reply:
Never tried it. As mentioned  in https://github.com/LibreSolar/charge-controller-firmware/pull/130/files#r866892157 I messed around a lot with this struct. 
NUMBER_OF_OBSERVABLES_GPS, NUMBER_OF_OBSERVABLES_GPS would need to become variable or I would need to redefine it, which I remember should not be done. 

Anyways the whole GPS Code is not needed long term. It is an example from TinyEFK library which I implemented as a test, to ensure that the library always behave correct, after code changes.